File: 2022/python/day_5.py
#!/usr/bin/env python3

import os
import copy

INPUT_FILEPATH = os.path.join(os.path.dirname(__file__), "../inputs/day_5.txt")


def parse_start_position(input):
    stacks = []
    for line in reversed(input[:-1]):
        pos = 1
        idx = 0
        while True:
            try:
                crate = line[pos]
                if len(stacks) <= idx:
                    stacks.append([])
                if crate != " ":
                    stacks[idx].append(crate)
                pos += 4
                idx += 1
            except IndexError:
                break
    return stacks


def main():

    start_position = []
    instructions = []
    with open(INPUT_FILEPATH) as input:
        while (line := input.readline()) != "\n":
            start_position.append(line)
        for line in input:
            raw_inst = line.strip().split(" ")
            instructions.append([int(raw_inst[1]), int(raw_inst[3])-1, int(raw_inst[5])-1 ])

    orig_stacks = parse_start_position(start_position)

    stacks = copy.deepcopy(orig_stacks)
    for inst in instructions:
        count, start, end = inst
        for _ in range(0, count):
            stacks[end].append(stacks[start].pop())
    result = ""
    for stack in stacks:
       result += stack[-1]
    print(result)

    stacks = copy.deepcopy(orig_stacks)
    for inst in instructions:
        count, start, end = inst
        moved_stack = []
        for _ in range(0, count):
            moved_stack.append(stacks[start].pop())
        stacks[end].extend(reversed(moved_stack))
    result = ""
    for stack in stacks:
       result += stack[-1]
    print(result)



if __name__ == "__main__":
    main()

File: 2022/rust/src/bin/day_5.rs
use std::fs::read_to_string;
use std::path::Path;

const INPUT_PATH_REL: &str = "../../../inputs/day_5.txt";
const SRC_PATH: &str = concat!(env!("CARGO_MANIFEST_DIR"), "/", file!());

fn parse_start_position(input: Vec<&str>) -> Vec<Vec<char>> {
    let mut stacks: Vec<Vec<char>> = vec![];
    for line in input.iter().rev().skip(1) {
        let mut pos = 1;
        let mut idx = 0;
        let chars = line.chars().collect::<Vec<_>>();
        loop {
            let element = match chars.get(pos) {
                None => break,
                Some(val) => *val,
            };
            if stacks.len() <= idx {
                stacks.push(vec![]);
            }
            if element != ' ' {
                stacks[idx].push(element);
            }
            pos += 4;
            idx += 1;
        }
    }
    stacks
}

fn parser_instructions(input: Vec<&str>) -> Vec<Instruction> {
    input
        .iter()
        .map(|s| s.split_whitespace().collect::<Vec<_>>())
        .map(|i| Instruction {
            count: i[1].parse().unwrap(),
            from: i[3].parse::<usize>().unwrap() - 1,
            to: i[5].parse::<usize>().unwrap() - 1,
        })
        .collect()
}

fn compute_result(stacks: &Vec<Vec<char>>) -> String {
    let mut result = String::new();
    for stack in stacks {
        let v = *stack.last().unwrap();
        result.push_str(&v.to_string());
    }
    result
}

struct Instruction {
    count: usize,
    from: usize,
    to: usize,
}

fn main() {
    let input_path = Path::new(SRC_PATH).parent().unwrap().join(INPUT_PATH_REL);
    let input = read_to_string(input_path).unwrap();

    let mut input_lines = input.lines();
    let input_lines_iter = input_lines.by_ref();
    let start_position_iter = input_lines_iter.take_while(|p| !p.is_empty());

    let orig_stacks = parse_start_position(start_position_iter.collect());
    let instructions = parser_instructions(input_lines_iter.collect());

    let mut stacks = orig_stacks.clone();
    for instruction in &instructions {
        for _ in 0..instruction.count {
            let elem = stacks[instruction.from].pop().unwrap();
            stacks[instruction.to].push(elem);
        }
    }
    println!("{}", compute_result(&stacks));

    stacks = orig_stacks;
    for instruction in &instructions {
        let mut elems = vec![];
        for _ in 0..instruction.count {
            let elem = stacks[instruction.from].pop().unwrap();
            elems.push(elem);
        }
        elems.reverse();
        stacks[instruction.to].extend(elems);
    }
    println!("{}", compute_result(&stacks));
}
